File: app/src/main/java/org/helllabs/android/xmp/Xmp.java
package org.helllabs.android.xmp;

import org.helllabs.android.xmp.util.ModInfo;

public final class Xmp {

    // Return codes
    public static final int XMP_END = 1;            // End of module reached

    // Sample format flags
    public static final int FORMAT_MONO = 1 << 2;

    // player parameters
    public static final int PLAYER_AMP = 0;            // Amplification factor
    public static final int PLAYER_MIX = 1;            // Stereo mixing
    public static final int PLAYER_INTERP = 2;        // Interpolation type
    public static final int PLAYER_DSP = 3;            // DSP effect flags
    public static final int PLAYER_CFLAGS = 5;      // Current module flags
    public static final int PLAYER_VOLUME = 7;        // Player volume (for audio focus duck)
    public static final int PLAYER_DEFPAN = 10;        // Default pan separation

    // Interpolation types
    public static final int INTERP_NEAREST = 0;        // Nearest neighbor
    public static final int INTERP_LINEAR = 1;        // Linear (default)
    public static final int INTERP_SPLINE = 2;        // Cubic spline

    // Player flags
    public static final int FLAGS_A500 = 1 << 3;

    // DSP effect types
    public static final int DSP_LOWPASS = 1 << 0;    // Lowpass filter effect

    // Limits
    public static final int MAX_CHANNELS = 64;        // Max number of channels in module


    private Xmp() {

    }

    public static native boolean init(int rate, int ms);

    public static native void deinit();

    public static native boolean testModule(String name, ModInfo info);

    public static native int loadModule(String name);

    public static native int releaseModule();

    public static native int startPlayer(int rate);

    public static native int endPlayer();

    public static native int playAudio();

    public static native int dropAudio();

    public static native boolean stopAudio();

    public static native boolean restartAudio();

    public static native boolean hasFreeBuffer();

    public static native int fillBuffer(boolean loop);

    public static native int nextPosition();

    public static native int prevPosition();

    public static native int setPosition(int num);

    public static native int stopModule();

    public static native int restartModule();

    public static native int seek(int time);

    public static native int time();

    public static native int mute(int chn, int status);

    public static native void getInfo(int[] values);

    public static native int getPlayer(int parm);

    public static native void setPlayer(int parm, int val);

    public static native int getLoopCount();

    public static native void getModVars(int[] vars);

    public static native String getVersion();

    public static native String getModName();

    public static native String getModType();

    public static native String getComment();

    public static native String[] getFormats();

    public static native String[] getInstruments();

    public static native void getChannelData(int[] volumes, int[] finalvols, int[] pans, int[] instruments, int[] keys, int[] periods);

    public static native void getPatternRow(int pat, int row, byte[] rowNotes, byte[] rowInstruments);

    public static native void getSampleData(boolean trigger, int ins, int key, int period, int chn, int width, byte[] buffer);

    public static native boolean setSequence(int seq);

    public static native void getSeqVars(int[] vars);

    public static native int getVolume();

    public static native int setVolume(int vol);

    static {
        System.loadLibrary("xmp-jni");
    }
}

File: app/src/main/java/org/helllabs/android/xmp/Xmp.kt
package org.helllabs.android.xmp

import org.helllabs.android.xmp.util.ModInfo

object Xmp {

    // Return codes
    // val XMP_END = 1                  // End of module reached

    // Sample format flags
    // val FORMAT_MONO = 1 shl 2

    // player parameters
    const val PLAYER_AMP = 0 // Amplification factor
    const val PLAYER_MIX = 1 // Stereo mixing
    const val PLAYER_INTERP = 2 // Interpolation type
    const val PLAYER_DSP = 3 // DSP effect flags
    const val PLAYER_CFLAGS = 5 // Current module flags
    // const val PLAYER_VOLUME = 7  // Player volume (for audio focus duck)
    const val PLAYER_DEFPAN = 10 // Default pan separation

    // Interpolation types
    const val INTERP_NEAREST = 0 // Nearest neighbor
    const val INTERP_LINEAR = 1 // Linear (default)
    const val INTERP_SPLINE = 2 // Cubic spline

    // Player flags
    const val FLAGS_A500 = 1 shl 3

    // DSP effect types
    const val DSP_LOWPASS = 1 shl 0 // Lowpass filter effect

    // Limits
    const val MAX_CHANNELS = 64 // Max number of channels in module

    init {
        System.loadLibrary("xmp")
    }

    external fun deinit(): Int
    external fun dropAudio()
    external fun endPlayer(): Int
    external fun fillBuffer(loop: Boolean): Int
    external fun getChannelData(
            volumes: IntArray,
            finalvols: IntArray,
            pans: IntArray,
            instruments: IntArray,
            keys: IntArray,
            periods: IntArray
    )
    external fun getComment(): String
    external fun getFormats(): Array<String>
    external fun getInfo(values: IntArray)
    external fun getInstruments(): Array<String>
    external fun getModName(): String
    external fun getModType(): String
    external fun getModVars(vars: IntArray)
    external fun getPatternRow(pat: Int, row: Int, rowNotes: ByteArray, rowInstruments: ByteArray)
    external fun getPlayer(parm: Int): Int
    external fun getSampleData(
            trigger: Boolean,
            ins: Int,
            key: Int,
            period: Int,
            chn: Int,
            width: Int,
            buffer: ByteArray
    )
    external fun getSeqVars(vars: IntArray)
    external fun getVersion(): String
    external fun getVolume(): Int
    external fun hasFreeBuffer(): Boolean
    external fun init(rate: Int, ms: Int): Boolean
    external fun loadModule(name: String): Int
    external fun mute(chn: Int, status: Int): Int
    external fun playAudio(): Int
    external fun releaseModule(): Int
    external fun restartAudio(): Boolean
    external fun seek(time: Int): Int
    external fun setPlayer(parm: Int, value: Int)
    external fun setSequence(seq: Int): Boolean
    external fun setVolume(vol: Int): Int
    external fun startPlayer(rate: Int): Int
    external fun stopAudio(): Boolean
    external fun stopModule(): Int
    external fun testModule(name: String, info: ModInfo): Boolean
    external fun time(): Int

    /*Not used, but still in JNI*/
//    external fun nextPosition(): Int
//    external fun prevPosition(): Int
//    external fun setPosition(num: Int): Int
//    external fun getLoopCount(): Int
//    external fun restartModule(): Int
}
